db/index.go

```diff
@@ -0,0 +1,495 @@
+			log.ErrorE(ctx, "Failed to close collection query", err)
+		}
+	}()
+
+	elements := make([]T, 0)
+	for res := range q.Next() {
+		if res.Error != nil {
+			return nil, res.Error
+		}
+
+		var element T
+		err = json.Unmarshal(res.Value, &element)
+		if err != nil {
+			return nil, NewErrInvalidStoredIndex(err)
+		}
+		elements = append(elements, element)
+	}
+	return elements, nil
+}
+
+func (c *collection) getIndexes(ctx context.Context, txn datastore.Txn) ([]CollectionIndex, error) {
+	if c.isIndexCached {
+		return c.indexes, nil
+	}
+
+	prefix := core.NewCollectionIndexKey(c.Name(), "")
+	if txn == nil {
+		var err error
+		txn, err = c.getTxn(ctx, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+	indexes, err := deserializePrefix[client.IndexDescription](ctx, prefix.ToString(), txn.Systemstore())
+	if err != nil {
+		return nil, err
+	}
+	colIndexes := make([]CollectionIndex, 0, len(indexes))
+	for _, index := range indexes {
+		colIndexes = append(colIndexes, NewCollectionIndex(c, index))
+	}
+
+	c.indexes = colIndexes
+	c.isIndexCached = true
+	return colIndexes, nil
+}
+
+func (c *collection) GetIndexes(ctx context.Context) ([]client.IndexDescription, error) {
+	indexes, err := c.getIndexes(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	indexDescriptions := make([]client.IndexDescription, 0, len(indexes))
+	for _, index := range indexes {
+		indexDescriptions = append(indexDescriptions, index.Description())
+	}
+
+	return indexDescriptions, nil
+}
+
+func (c *collection) createIndex(
+	ctx context.Context,
+	desc client.IndexDescription,
+) (CollectionIndex, error) {
+	err := validateIndexDescription(desc)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.checkExistingFields(ctx, desc.Fields)
+	if err != nil {
+		return nil, err
+	}
+
+	indexKey, err := c.processIndexName(ctx, &desc)
+	if err != nil {
+		return nil, err
+	}
+
+	txn, err := c.getTxn(ctx, false)
+	if err != nil {
+		return nil, err
+	}
+
+	colSeq, err := c.db.getSequence(ctx, txn, fmt.Sprintf("%s/%d", core.COLLECTION_INDEX, c.ID()))
+	if err != nil {
+		return nil, err
+	}
+	colID, err := colSeq.next(ctx, txn)
+	if err != nil {
+		return nil, err
+	}
+	desc.ID = uint32(colID)
+
+	buf, err := json.Marshal(desc)
+	if err != nil {
+		return nil, err
+	}
+
+	err = txn.Systemstore().Put(ctx, indexKey.ToDS(), buf)
+	if err != nil {
+		return nil, err
+	}
+	colIndex := NewCollectionIndex(c, desc)
+	return colIndex, nil
+}
+
+func (c *collection) checkExistingFields(
+	ctx context.Context,
+	fields []client.IndexedFieldDescription,
+) error {
+	collectionFields := c.Description().Schema.Fields
+	for _, field := range fields {
+		found := false
+		fieldLower := strings.ToLower(field.Name)
+		for _, colField := range collectionFields {
+			if fieldLower == strings.ToLower(colField.Name) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return NewErrNonExistingFieldForIndex(field.Name)
+		}
+	}
+	return nil
+}
+
+func (c *collection) processIndexName(
```

Comment:
suggestion: Can you please rename of document this function as the name doesn't make it clear as to what is happening bellow.